frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/tolerations-modal/tolerations-modal.tsx

```diff
@@ -0,0 +1,166 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+import { ModalTitle, ModalBody, ModalComponentProps } from '@console/internal/components/factory';
+import {
+  Button,
+  ButtonVariant,
+  GridItem,
+  FormSelect,
+  FormSelectOption,
+} from '@patternfly/react-core';
+import {
+  FirehoseResult,
+  withHandlePromise,
+  HandlePromiseProps,
+} from '@console/internal/components/utils';
+import { k8sPatch, Toleration, TaintEffect, K8sResourceKind } from '@console/internal/module/k8s';
+import { NodeModel } from '@console/internal/models';
+import { isLoaded, getLoadedData, getLoadError } from '../../../../utils';
+import { ModalFooter } from '../../modal/modal-footer';
+import { VMLikeEntityKind } from '../../../../types/vmLike';
+import { getVMLikeModel, getTolerations } from '../../../../selectors/vm';
+import { getTolerationsPatch } from '../../../../k8s/patches/vm/vm-scheduling-patches';
+import { NodeChecker } from '../shared/NodeChecker/node-checker';
+import { useNodeQualifier } from '../shared/hooks';
+import { LabelsList } from '../../../LabelsList/labels-list';
+import { TOLERATIONS_MODAL_TITLE, TOLERATIONS_EFFECTS } from '../shared/consts';
+import { useIDEntities } from '../../../../hooks/use-id-entities';
+import { useCollisionChecker } from '../../../../hooks/use-collision-checker';
+import { IDLabel } from '../../../LabelsList/types';
+
+export const TModal = withHandlePromise(
+  ({
+    nodes,
+    close,
+    handlePromise,
+    inProgress,
+    errorMessage,
+    vmLikeEntity,
+    vmLikeEntityLoading,
+  }: TModalProps) => {
+    const vmLikeFinal = getLoadedData(vmLikeEntityLoading, vmLikeEntity);
+    const loadError = getLoadError(nodes, NodeModel);
+
+    const [
+      tolerationsLabels,
+      setTolerationsLabels,
+      onLabelAdd,
+      onLabelChange,
+      onLabelDelete,
+    ] = useIDEntities<TolerationLabel>(
+      getTolerations(vmLikeEntity)?.map((toleration, id) => ({ ...toleration, id })) || [],
+    );
+
+    const qualifiedNodes = useNodeQualifier(tolerationsLabels, nodes);
+
+    const [showCollisionAlert, reload] = useCollisionChecker<VMLikeEntityKind>(
+      vmLikeFinal,
+      (oldVM: VMLikeEntityKind, newVM: VMLikeEntityKind) =>
+        _.isEqual(getTolerations(oldVM), getTolerations(newVM)),
+    );
+
+    const onTolerationAdd = () =>
+      onLabelAdd({
+        id: null,
+        key: '',
+        value: '',
+        effect: TOLERATIONS_EFFECTS[0],
+      } as TolerationLabel);
+
+    const onReload = () => {
+      reload();
+      setTolerationsLabels(
+        getTolerations(vmLikeFinal)?.map((toleration, id) => ({
+          ...toleration,
+          id,
+        })) || [],
```

Comment:
same here

Reply:
we should keep this one, if the selector returns undefined.

Reply:
ah, true